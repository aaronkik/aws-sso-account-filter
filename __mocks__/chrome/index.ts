import { ChromeStorageArea } from './chrome-storage';

export const chrome = { storage: { sync: new ChromeStorageArea() } };
