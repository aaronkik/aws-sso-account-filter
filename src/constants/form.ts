export const ACCOUNT_FILTER_REQUIRED_MESSAGE = 'Filter cannot be empty';
