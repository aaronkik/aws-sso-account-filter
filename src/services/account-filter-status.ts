import { ChromeStorageSync } from '../repositories';

const accountFiltersStatusKey = 'accountFilterStatus';

type GetAccountFilterStatus = {
  [accountFiltersStatusKey]?: boolean;
};

class AccountFilterStatus {
  #storageKey = accountFiltersStatusKey;
  #storage = new ChromeStorageSync();

  async #set(status: boolean) {
    return this.#storage.set({ accountFilterStatus: status });
  }

  async get() {
    return new Promise<boolean>((resolve, reject) => {
      this.#storage
        .get<GetAccountFilterStatus>(this.#storageKey)
        .then((result) => {
          if (
            result &&
            'accountFilterStatus' in result &&
            typeof result.accountFilterStatus === 'boolean'
          ) {
            resolve(result.accountFilterStatus);
          } else {
            this.enable();
            resolve(true);
          }
        })
        .catch((error) => {
          reject(error);
        });
    });
  }

  async enable() {
    return this.#set(true);
  }

  async disable() {
    return this.#set(false);
  }
}

export const accountFilterStatus = new AccountFilterStatus();
