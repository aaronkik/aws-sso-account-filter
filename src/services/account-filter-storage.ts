import { ChromeStorageSync } from '~/repositories';

const accountFiltersStorageKey = 'accountFilters';

type AccountFilter = {
  id: string;
  filter: string;
};

type GetAccountFilterStorage = {
  [accountFiltersStorageKey]?: Array<AccountFilter>;
};

export class AccountFilterStorage {
  #storageKey = accountFiltersStorageKey;
  #storage = new ChromeStorageSync();

  async get() {
    return this.#storage.get<GetAccountFilterStorage>(this.#storageKey);
  }

  async set(accountFilters: Array<AccountFilter>) {
    return this.#storage.set({ accountFilters });
  }
}
