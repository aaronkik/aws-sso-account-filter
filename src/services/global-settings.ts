import { ChromeStorageSync } from '../repositories';

interface GetGlobalSettings {
  accountFilterStatus?: boolean;
  autoRefreshPage?: boolean;
}

export type GlobalSettingKey = keyof GetGlobalSettings;

class GlobalSettings extends ChromeStorageSync {
  async getValue(key: GlobalSettingKey): Promise<boolean> {
    const result = await super.get<GetGlobalSettings>(key);

    if (result && key in result && typeof result[key] === 'boolean') {
      return result[key];
    }

    await super.set({ [key]: true });
    return true;
  }

  async setValue(key: GlobalSettingKey, value: boolean): Promise<void> {
    return super.set({ [key]: value });
  }
}

export const globalSettings = new GlobalSettings();
