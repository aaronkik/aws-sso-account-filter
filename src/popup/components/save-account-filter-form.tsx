import { useForm } from 'react-hook-form';
import { Button, FormErrorMessage, Input } from '../../components';
import { ACCOUNT_FILTER_REQUIRED_MESSAGE } from '../../constants/form';
import { accountFilterStorage } from '../../services/account-filter-storage';

type FormValues = {
  accountFilter: string;
};

const SaveAccountFilterForm = () => {
  const {
    formState: { errors },
    handleSubmit,
    register,
    reset,
  } = useForm<FormValues>({
    defaultValues: {
      accountFilter: '',
    },
  });

  const saveAccountFilter = handleSubmit(async ({ accountFilter }) => {
    try {
      await accountFilterStorage.add(accountFilter);
      reset({ accountFilter: '' });
    } catch (error) {
      console.error(error);
    }
  });

  return (
    <form
      autoComplete='off'
      aria-label='Save account filter form'
      id='save-account-filter-form'
      onSubmit={saveAccountFilter}
    >
      <div className='flex gap-4'>
        <div className='relative w-full'>
          <div className='pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3'>
            <span className='text-lg font-medium text-slate-500'>/</span>
          </div>
          <Input
            aria-label='Account filter input'
            className='px-6'
            inputMode='text'
            placeholder='team-(sandbox|dev)'
            type='text'
            {...register('accountFilter', {
              required: ACCOUNT_FILTER_REQUIRED_MESSAGE,
              validate: {
                isValidRegex: (value) => {
                  try {
                    new RegExp(value, 'u');
                    return true;
                  } catch (error) {
                    if (error instanceof SyntaxError) {
                      return error.message;
                    }
                    return `Unknown RegExp error: value=${value}`;
                  }
                },
              },
            })}
          />
          <div className='pointer-events-none absolute inset-y-0 right-0 flex items-center pr-3'>
            <span className='text-lg font-medium text-slate-500'>/</span>
          </div>
        </div>
        <Button aria-label='Save account filter' className='min-w-[6rem]' type='submit'>
          Save
        </Button>
      </div>
      {errors.accountFilter && (
        <FormErrorMessage className='mt-2'>{errors.accountFilter.message}</FormErrorMessage>
      )}
    </form>
  );
};

export default SaveAccountFilterForm;
