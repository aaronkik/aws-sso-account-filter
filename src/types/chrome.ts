import { type AccountFilter } from '../services/account-filter-storage';
import { accountFilterStatus } from '../services/account-filter-status';
import { type GlobalSettingKey, globalSettings } from '../services/global-settings';

interface GenericChromeStorageChange<Type> extends chrome.storage.StorageChange {
  newValue?: Type;
  oldValue?: Type;
}

export type AccountFilterChromeStorageChange = Record<
  'accountFilters',
  GenericChromeStorageChange<Array<AccountFilter>>
>;

export type AccountFilterStatusChromeStorageChange = Record<
  'accountFilterStatus',
  GenericChromeStorageChange<Awaited<ReturnType<typeof accountFilterStatus.get>>>
>;

export type GlobalSettingsChromeStorageChange = Record<
  GlobalSettingKey,
  GenericChromeStorageChange<Awaited<ReturnType<typeof globalSettings.getValue>>>
>;
