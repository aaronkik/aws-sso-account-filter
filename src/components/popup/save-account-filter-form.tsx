import { useForm } from 'react-hook-form';
import { Input, Button, FormErrorMessage } from '~/components/shared';
import { ACCOUNT_FILTER_REQUIRED_MESSAGE } from '~/constants/form';
import { AccountFilterStorage } from '~/services/account-filter-storage';

type FormValues = {
  accountFilter: string;
};

const accountFilterStorage = new AccountFilterStorage();

const SaveAccountFilterForm = () => {
  const {
    formState: { errors },
    handleSubmit,
    register,
    reset,
  } = useForm<FormValues>({
    defaultValues: {
      accountFilter: '',
    },
  });

  const saveAccountFilter = handleSubmit(async ({ accountFilter }) => {
    try {
      await accountFilterStorage.add(accountFilter);
      reset({ accountFilter: '' });
    } catch (error) {
      console.error(error);
    }
  });

  return (
    <form
      autoComplete='off'
      aria-label='Save account filter form'
      id='save-account-filter-form'
      onSubmit={saveAccountFilter}
    >
      <div className='flex gap-4'>
        <Input
          aria-label='Account filter input'
          inputMode='text'
          placeholder='dev'
          type='text'
          {...register('accountFilter', { required: ACCOUNT_FILTER_REQUIRED_MESSAGE })}
        />
        <Button aria-label='Save account filter' className='min-w-[8rem]' type='submit'>
          Save filter
        </Button>
      </div>
      {errors.accountFilter && (
        <FormErrorMessage className='mt-2'>{errors.accountFilter.message}</FormErrorMessage>
      )}
    </form>
  );
};

export default SaveAccountFilterForm;
