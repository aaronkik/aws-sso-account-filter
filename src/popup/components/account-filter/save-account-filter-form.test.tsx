import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { vi, describe, expect, test } from 'vitest';
import { ACCOUNT_FILTER_REQUIRED_MESSAGE } from '../../../constants/form';
import { accountFilterStorage } from '../../../services/account-filter-storage';
import { SaveAccountFilterForm } from './save-account-filter-form';
import { chrome } from '../../../../__mocks__/chrome';

const initialSetup = () => {
  const user = userEvent.setup();
  render(<SaveAccountFilterForm />);
  const accountFilterInput = screen.getByLabelText('Account filter input');
  const saveFilterButton = screen.getByRole('button', { name: /Save account filter/i });
  return { accountFilterInput, saveFilterButton, user };
};

describe('SaveAccountFilterForm', () => {
  test('matches snapshot', () => {
    const { container } = render(<SaveAccountFilterForm />);
    expect(container).toMatchSnapshot();
  });

  test('renders an error message when the account filter input is empty and the save filter button submits', async () => {
    const { saveFilterButton, user } = initialSetup();

    expect(screen.queryByText(ACCOUNT_FILTER_REQUIRED_MESSAGE)).not.toBeInTheDocument();

    await user.click(saveFilterButton);

    expect(screen.getByText(ACCOUNT_FILTER_REQUIRED_MESSAGE)).toBeInTheDocument();
  });

  test('chrome storage does not add a new filter when save filter button submits and the account filter input is empty', async () => {
    const { saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    const chromeStorageSetSpy = vi.spyOn(chrome.storage.sync, 'set');

    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).not.toHaveBeenCalled();
    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });
  });

  test('chrome storage adds the user typed input as a new filter when save filter button submits', async () => {
    const { accountFilterInput, saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    const chromeStorageSetSpy = vi.spyOn(chrome.storage.sync, 'set');

    await user.type(accountFilterInput, 'test');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toHaveBeenCalledWith({
      accountFilters: expect.arrayContaining([expect.objectContaining({ filter: 'test' })]),
    });
    expect(chromeStorageSetSpy).toBeCalledTimes(1);

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({
      accountFilters: expect.arrayContaining([expect.objectContaining({ filter: 'test' })]),
    });
  });

  test('account filter input is cleared after a user has typed an input as a new filter and the save filter button submits', async () => {
    const { accountFilterInput, saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    expect(accountFilterInput).toHaveValue('');

    await user.type(accountFilterInput, 'test');
    expect(accountFilterInput).toHaveValue('test');

    await user.click(saveFilterButton);

    await waitFor(() => {
      expect(accountFilterInput).toHaveValue('');
    });
  });

  test('does not add duplicate filter names', async () => {
    const { accountFilterInput, saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    const chromeStorageSetSpy = vi.spyOn(chrome.storage.sync, 'set');

    await user.type(accountFilterInput, 'test');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toBeCalledTimes(1);
    await expect(accountFilterStorage.get()).resolves.toStrictEqual({
      accountFilters: expect.arrayContaining([expect.objectContaining({ filter: 'test' })]),
    });

    await user.type(accountFilterInput, 'test');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toBeCalledTimes(2);
    await expect(accountFilterStorage.get()).resolves.toStrictEqual({
      accountFilters: expect.arrayContaining([expect.objectContaining({ filter: 'test' })]),
    });
  });

  test('adds multiple filter names', async () => {
    const { accountFilterInput, saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    const chromeStorageSetSpy = vi.spyOn(chrome.storage.sync, 'set');

    await user.type(accountFilterInput, 'foo');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toBeCalledTimes(1);
    await expect(accountFilterStorage.get()).resolves.toEqual({
      accountFilters: [expect.objectContaining({ filter: 'foo' })],
    });

    await user.type(accountFilterInput, 'bar');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toBeCalledTimes(2);
    await expect(accountFilterStorage.get()).resolves.toEqual({
      accountFilters: [expect.objectContaining({ filter: 'bar' }), expect.objectContaining({ filter: 'foo' })],
    });
  });

  test('removes whitespace from user filter input before saving to storage', async () => {
    const { accountFilterInput, saveFilterButton, user } = initialSetup();

    await expect(accountFilterStorage.get()).resolves.toStrictEqual({ accountFilters: [] });

    const chromeStorageSetSpy = vi.spyOn(chrome.storage.sync, 'set');

    await user.type(accountFilterInput, '  foo  ');
    await user.click(saveFilterButton);

    expect(chromeStorageSetSpy).toBeCalledTimes(1);
    await expect(accountFilterStorage.get()).resolves.toEqual({
      accountFilters: [expect.objectContaining({ filter: 'foo' })],
    });
  });
});
