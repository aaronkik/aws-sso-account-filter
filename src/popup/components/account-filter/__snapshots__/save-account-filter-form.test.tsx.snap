// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SaveAccountFilterForm > matches snapshot 1`] = `
<div>
  <form
    aria-label="Save account filter form"
    autocomplete="off"
    id="save-account-filter-form"
  >
    <div
      class="flex gap-2"
    >
      <div
        class="relative w-full"
      >
        <div
          class="pointer-events-none absolute inset-y-0 left-0 flex items-center pl-3"
        >
          <span
            class="text-lg font-medium text-slate-500"
          >
            /
          </span>
        </div>
        <input
          aria-label="Account filter input"
          class="w-full rounded-sm border border-transparent bg-slate-800 leading-6 placeholder:text-slate-500 focus:border-orange-400 focus:bg-transparent focus:ring-orange-400 focus:placeholder:text-slate-400 px-6"
          inputmode="text"
          name="accountFilter"
          placeholder="team-(sandbox|dev)"
          type="text"
        />
        <div
          class="pointer-events-none absolute inset-y-0 right-0 flex items-center pr-3"
        >
          <span
            class="text-lg font-medium text-slate-500"
          >
            /
          </span>
        </div>
      </div>
      <button
        aria-label="Save account filter"
        class="inline-flex items-center justify-center rounded-sm bg-orange-400 px-4 py-1.5 text-base leading-6 font-medium tracking-wider transition-colors duration-150 select-none text-shadow-lg hover:bg-orange-500 focus-visible:bg-orange-600 active:bg-orange-600 min-w-20"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</div>
`;
